src/ssl.c

```diff
@@ -15924,198 +15954,150 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
-
-                ret = wolfSSL_sk_X509_NAME_new(NULL);
-                do {
-                    x509 = wolfSSL_sk_X509_pop(sk);
-                    if (x509 != NULL) {
-                        if (wolfSSL_X509_get_isCA(x509)) {
-                            if (wolfSSL_sk_X509_NAME_push(ret,
-                                    wolfSSL_X509_get_subject_name(x509)) != 0) {
-                                WOLFSSL_MSG("Error pushing X509 name to stack");
-                                /* continue on to try other certificates and
-                                 * do not fail out here */
-                            }
-                        }
-                        wolfSSL_X509_free(x509);
-                    }
-                } while (x509 != NULL);
-                wolfSSL_sk_X509_free(sk);
-                return ret;
-            }
-            return NULL;
-        }
-        else {
-            /* currently only can be set in the CTX */
-            return ssl->ctx->ca_names;
-        }
+        return SSL_CA_NAMES(ssl);
     }
-#endif /* SESSION_CERTS */
 
-
-    #if defined(OPENSSL_ALL) || defined(OPENSSL_EXTRA) || \
-        defined(WOLFSSL_NGINX) || defined (WOLFSSL_HAPROXY)
-    /* registers client cert callback, called during handshake if server
-       requests client auth but user has not loaded client cert/key */
-    void wolfSSL_CTX_set_client_cert_cb(WOLFSSL_CTX *ctx, client_cert_cb cb)
+    #if !defined(NO_RSA) && !defined(NO_CERTS)
+    int wolfSSL_CTX_add_client_CA(WOLFSSL_CTX* ctx, WOLFSSL_X509* x509)
     {
-        WOLFSSL_ENTER("wolfSSL_CTX_set_client_cert_cb");
+        WOLFSSL_X509_NAME *nameCopy = NULL;
 
-        if (ctx != NULL) {
-            ctx->CBClientCert = cb;
+        WOLFSSL_ENTER("wolfSSL_CTX_add_client_CA");
+
+        if (ctx == NULL || x509 == NULL){
+            WOLFSSL_MSG("Bad argument");
+            return WOLFSSL_FAILURE;
         }
-    }
-    #endif /* OPENSSL_ALL || OPENSSL_EXTRA || WOLFSSL_NGINX || WOLFSSL_HAPROXY */
 
-#endif /* OPENSSL_EXTRA || WOLFSSL_EXTRA || HAVE_WEBSERVER */
+        if (ctx->ca_names == NULL) {
+            ctx->ca_names = wolfSSL_sk_X509_NAME_new(NULL);
+            if (ctx->ca_names == NULL) {
+                WOLFSSL_MSG("wolfSSL_sk_X509_NAME_new error");
+                return WOLFSSL_FAILURE;
+            }
+        }
 
-#if defined(OPENSSL_EXTRA) || defined(WOLFSSL_EXTRA)
-    WOLF_STACK_OF(WOLFSSL_X509_NAME)* wolfSSL_CTX_get_client_CA_list(
-            const WOLFSSL_CTX *s)
-    {
-        WOLFSSL_ENTER("wolfSSL_CTX_get_client_CA_list");
+        nameCopy = wolfSSL_X509_NAME_dup(wolfSSL_X509_get_subject_name(x509));
+        if (nameCopy == NULL) {
+            WOLFSSL_MSG("wolfSSL_X509_NAME_dup error");
+            return WOLFSSL_FAILURE;
+        }
 
-        if (s == NULL)
-            return NULL;
+        if (wolfSSL_sk_X509_NAME_push(ctx->ca_names, nameCopy) != WOLFSSL_SUCCESS) {
+            WOLFSSL_MSG("wolfSSL_sk_X509_NAME_push error");
```

Comment:
In this error case, should `ctx->ca_names` be freed if it was allocated in this function call? I.e. if the conditional at line 15972 was taken.

Reply:
`ctx->ca_names` will be `free`'d when `free`'ing the entire `WOLFSSL_CTX` structure. We don't need to do it here.